### PR TITLE
Fixed #35051 -- Prevented runserver from removing non-zero Content-Length for HEAD requests.

### PR DESCRIPTION
**Pull request by pizzapanther**
Adds a check to only delete Content-Length header if it is zero. This allows implementation of protocols like [HTTP Range Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) where Content-Length is required in the HEAD request.

Reference: https://forum.djangoproject.com/t/optionally-do-not-drop-content-length-for-head-requests/26305/1
Ticket: https://code.djangoproject.com/ticket/35051

Also updated the `.gitignore` to include the standard python venv path so people don't accidentally commit it.

### CODE REVIEW

**Comment by HamaBarhamou** [comment]
> still working on it. I'll remove `Draft:` when its ready.

yes indeed you're right. i should have waited until everything was ready for a revision. it slipped my mind. good work.

**Comment by pizzapanther** [comment]
Ready for review. Merry Christmas! 🎄🎄🎄

CC: @HamaBarhamou

**Comment by HamaBarhamou** [comment]
> Ready for review. Merry Christmas! 🎄🎄🎄
> 
> CC: @HamaBarhamou

hi, merry christmas to you too. your tiket doesn't appear in the tiket ready for revision [list](https://code.djangoproject.com/query?status=!closed&needs_better_patch=0&needs_tests=0&needs_docs=0&has_patch=1&stage=Accepted&desc=1&order=changetime). make sure you mention your path in the [tiket](https://code.djangoproject.com/ticket/35051) and check the box with a path. then check that it appears in the tiket ready for revision [list](https://code.djangoproject.com/query?status=!closed&needs_better_patch=0&needs_tests=0&needs_docs=0&has_patch=1&stage=Accepted&desc=1&order=changetime).

**Comment by pizzapanther** [comment]
thanks @HamaBarhamou updated

File: .gitignore [hunk 1]
```diff
@@ -6,6 +6,7 @@
 *.pot
 *.py[co]
 .tox/
+.venv
```
Comment (apollo13) [inline .gitignore:9]: Please don't include unrelated changes.

File: django/core/servers/basehttp.py [hunk 2]
```diff
@@ -134,6 +134,7 @@ def cleanup_headers(self):
         if (
             self.environ["REQUEST_METHOD"] == "HEAD"
             and "Content-Length" in self.headers
+            and str(self.headers["Content-Length"]) == "0"
```
Comment (apollo13) [inline django/core/servers/basehttp.py:137]: Wouldn't it be technically valid for a 200 response to return a `Content-Length` of `0`?

File: .gitignore (hunk 1 above)
Comment (pizzapanther) [reply .gitignore]: removed

File: django/core/servers/basehttp.py (hunk 2 above)
Comment (pizzapanther) [reply django/core/servers/basehttp.py:137]: HEAD request Header and subsequent GET request headers should match. So the only time you could have `Content-Length: 0` is if your GET request has an empty body. This is not very likely in general and less likely if you are using a HEAD request since they are most often used in situations where you have a large response.

In a perfect world, there would be a mechanism to see if the user set the content-length. However, checking for 0 is pretty much a good proxy for that without building out a specific mechanism for it. At least that is what I'm thinking.

**Comment by HamaBarhamou** [comment]
@pizzapanther  if you feel you have finished and are ready for a new review. uncheck this 'Patch needs improvement' box in your tiket and save, so your tiket will reappear in the list of tiket ready for a review.

File: django/core/servers/basehttp.py (hunk 2 above)
Comment (apollo13) [reply django/core/servers/basehttp.py:137]: Yeah, given that this is only for runserver,, I think it wouldn't be worth to add an extra mechanism for it.

File: tests/servers/tests.py [hunk 3]
```diff
@@ -256,6 +256,25 @@ def test_keep_alive_on_connection_with_content_length(self):
...
+        conn.close()
+
+    def test_head_content_length_set(self):
+        conn = HTTPConnection(
+            LiveServerViews.server_thread.host, LiveServerViews.server_thread.port
+        )
+        conn.request("HEAD", "/head_view/?content-length=11151978")
+        response = conn.getresponse()
+        self.assertEqual(response.getheader("Content-Length"), "11151978")
+        conn.close()
```
Comment (felixxm) [inline tests/servers/tests.py:276]: Move tests to the `tests/servers/test_basehttp.py`.

Comment (pizzapanther) [reply tests/servers/tests.py]: moved

File: tests/servers/test_basehttp.py [hunk 4]
```diff
@@ -189,3 +190,27 @@ def test_broken_pipe_errors(self):
...
+
+
+@override_settings(ROOT_URLCONF="servers.urls")
+class LiveServerViews(LiveServerTestCase):
+    available_apps = []
+
+    def test_head_content_length_removed(self):
+        conn = HTTPConnection(
+            LiveServerViews.server_thread.host, LiveServerViews.server_thread.port
+        )
```
Comment (felixxm) [inline tests/servers/test_basehttp.py:202]: Do we need to use `LiveServer` in tests? :thinking:

Comment (pizzapanther) [reply tests/servers/test_basehttp.py:202]: `ServerHandler` where the change was made is used by `runserver` (`WSGIServer`) and `LiveServer`. I couldn't find any tests that directly use `WSGIServer` and make HTTP calls the were similar to what I wanted but I did find the `LiveServer` calls. That's why I stuck it originally where I had it. Those `LiveServer` tests are basically how I wanted to test it.

So we have to use `LiveServer` or `WSGIServer` or maybe some kind of unit test on `ServerHandler`. I went the with the path that had tests already that I could base mine off of. I could probably figure out testing `ServerHandler` or `WSGIServer` but it was a longer task t figure out and `LiveServer` tested the code just fine.

Let me know your recommendations.

File: tests/servers/test_basehttp.py [hunk 5]
```diff
@@ -189,3 +190,27 @@ def test_broken_pipe_errors(self):
                         self.assertEqual(cm.records[0].getMessage(), msg)
                 finally:
                     server.server_close()
+
+
+@override_settings(ROOT_URLCONF="servers.urls")
+class LiveServerViews(LiveServerTestCase):
+    available_apps = []
+
+    def test_head_content_length_removed(self):
```
Comment (felixxm) [inline tests/servers/test_basehttp.py:199]: This is already covered by `servers.test_basehttp.WSGIRequestHandlerTestCase.test_no_body_returned_for_head_requests`.

File: tests/servers/test_basehttp.py [hunk 6]
```diff
@@ -189,3 +190,27 @@ def test_broken_pipe_errors(self):
...
+        conn.close()
+
+    def test_head_content_length_set(self):
+        conn = HTTPConnection(
+            LiveServerViews.server_thread.host, LiveServerViews.server_thread.port
+        )
+        conn.request("HEAD", "/head_view/?content-length=11151978")
+        response = conn.getresponse()
+        self.assertEqual(response.getheader("Content-Length"), "11151978")
+        conn.close()
```
Comment (felixxm) [inline tests/servers/test_basehttp.py:216]: We could test this with `WSGIRequestHandler`, e.g.
```diff
diff --git a/tests/servers/test_basehttp.py b/tests/servers/test_basehttp.py
index 1e535e933e..cc4701114a 100644
--- a/tests/servers/test_basehttp.py
+++ b/tests/servers/test_basehttp.py
@@ -161,6 +161,45 @@ class WSGIRequestHandlerTestCase(SimpleTestCase):
         )
         self.assertNotIn(b"Connection: close\r\n", lines)
 
+    def test_non_zero_content_length_set_head_request(self):
+        hello_world_body = b"<!DOCTYPE html><html><body>Hello World</body></html>"
+        content_length = len(hello_world_body)
+
+        def test_app(environ, start_response):
+            """
+            A WSGI app that returns a hello world with non-zero Content-Length.
+            """
+            start_response("200 OK", [("Content-length", str(content_length))])
+            return [hello_world_body]
+
+        rfile = BytesIO(b"HEAD / HTTP/1.0\r\n")
+        rfile.seek(0)
+
+        wfile = UnclosableBytesIO()
+
+        def makefile(mode, *a, **kw):
+            if mode == "rb":
+                return rfile
+            elif mode == "wb":
+                return wfile
+
+        request = Stub(makefile=makefile)
+        server = Stub(base_environ={}, get_app=lambda: test_app)
+
+        # Prevent logging from appearing in test output.
+        with self.assertLogs("django.server", "INFO"):
+            # Instantiating a handler runs the request as side effect.
+            WSGIRequestHandler(request, "192.168.0.2", server)
+
+        wfile.seek(0)
+        lines = list(wfile.readlines())
+        body = lines[-1]
+        # The body is not returned in a HEAD response.
+        self.assertEqual(body, b"\r\n")
+        # Non-zero Content-Length is not removed.
+        self.assertEqual(lines[-2], f"Content-length: {content_length}\r\n".encode())
+        self.assertNotIn(b"Connection: close\r\n", lines)
+
 
 class WSGIServerTestCase(SimpleTestCase):
     request_factory = RequestFactory()

```

File: tests/servers/test_basehttp.py (hunk 4 above)
Comment (felixxm) [reply tests/servers/test_basehttp.py:202]: `test_head_content_length_removed` scenario is already covered by `servers.test_basehttp.WSGIRequestHandlerTestCase.test_no_body_returned_for_head_requests`, so it should be enough to add a new `WSGIRequestHandler` test (see https://github.com/django/django/pull/17643#discussion_r1437997140).

File: tests/servers/test_basehttp.py (hunk 5 above)
Comment (pizzapanther) [reply tests/servers/test_basehttp.py]: tests updated

File: tests/servers/test_basehttp.py (hunk 6 above)
Comment (pizzapanther) [reply tests/servers/test_basehttp.py]: tests updated

**Comment by felixxm** [comment]
@pizzapanther Thanks :+1: Welcome aboard :boat:

**Comment by bigsbug** [comment]
hey @pizzapanther I appreciate your Fix!
